File: csharp/tests/ExponentialBackoffRetryPolicyTest.cs
/*
 * Licensed to the Apache Software Foundation (ASF) under one or more
 * contributor license agreements.  See the NOTICE file distributed with
 * this work for additional information regarding copyright ownership.
 * The ASF licenses this file to You under the Apache License, Version 2.0
 * (the "License"); you may not use this file except in compliance with
 * the License.  You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

using System;
using Apache.Rocketmq.V2;
using Google.Protobuf.WellKnownTypes;
using Microsoft.VisualStudio.TestTools.UnitTesting;
using Org.Apache.Rocketmq;

namespace tests
{
    [TestClass]
    public class ExponentialBackoffRetryPolicyTest
    {
        private TimeSpan initialBackoff = TimeSpan.FromMilliseconds(5);
        private TimeSpan maxBackoff = TimeSpan.FromSeconds(1);
        private float backoffMultiplier = 5;
        private int maxAttempts = 3;

        [TestMethod]
        public void TestNextAttemptDelayForImmediatelyRetryPolicy()
        {
            var retryPolicy = ExponentialBackoffRetryPolicy.ImmediatelyRetryPolicy(3);
            for (int i = 1; i <= 4; i++)
            {
                Assert.AreEqual(TimeSpan.Zero, retryPolicy.GetNextAttemptDelay(i));
            }
        }

        [TestMethod]
        [ExpectedException(typeof(ArgumentException))]
        public void TestGetNextAttemptDelayWithIllegalAttempt()
        {
            var retryPolicy = new ExponentialBackoffRetryPolicy(maxAttempts, initialBackoff, maxBackoff, backoffMultiplier);
            retryPolicy.GetNextAttemptDelay(0);
        }

        [TestMethod]
        public void TestGetNextAttemptDelay()
        {
            var retryPolicy = new ExponentialBackoffRetryPolicy(maxAttempts, initialBackoff, maxBackoff, backoffMultiplier);
            Assert.AreEqual(TimeSpan.FromMilliseconds(5), retryPolicy.GetNextAttemptDelay(1));
            Assert.AreEqual(TimeSpan.FromMilliseconds(25), retryPolicy.GetNextAttemptDelay(2));
            Assert.AreEqual(TimeSpan.FromMilliseconds(125), retryPolicy.GetNextAttemptDelay(3));
            Assert.AreEqual(TimeSpan.FromMilliseconds(625), retryPolicy.GetNextAttemptDelay(4));
            Assert.AreEqual(TimeSpan.FromSeconds(1), retryPolicy.GetNextAttemptDelay(5));
        }

        [TestMethod]
        public void TestFromProtobuf()
        {
            var initialBackoffProto = Duration.FromTimeSpan(initialBackoff);
            var maxBackoffProto = Duration.FromTimeSpan(maxBackoff);

            var exponentialBackoff = new ExponentialBackoff
            {
                Initial = initialBackoffProto,
                Max = maxBackoffProto,
                Multiplier = backoffMultiplier
            };
            var retryPolicyProto = new RetryPolicy
            {
                MaxAttempts = maxAttempts,
                ExponentialBackoff = exponentialBackoff
            };

            var policy = ExponentialBackoffRetryPolicy.FromProtobuf(retryPolicyProto);

            Assert.AreEqual(maxAttempts, policy.GetMaxAttempts());
            Assert.AreEqual(initialBackoff, policy.InitialBackoff);
            Assert.AreEqual(maxBackoff, policy.MaxBackoff);
            Assert.AreEqual(backoffMultiplier, policy.BackoffMultiplier);
        }

        [TestMethod]
        [ExpectedException(typeof(ArgumentException))]
        public void TestFromProtobufWithoutExponentialBackoff()
        {
            var customizedBackoff = new CustomizedBackoff();
            var retryPolicyProto = new RetryPolicy
            {
                MaxAttempts = maxAttempts,
                CustomizedBackoff = customizedBackoff
            };
            ExponentialBackoffRetryPolicy.FromProtobuf(retryPolicyProto);
        }

        [TestMethod]
        public void TestToProtobuf()
        {
            var retryPolicy = new ExponentialBackoffRetryPolicy(maxAttempts, initialBackoff, maxBackoff, backoffMultiplier);
            var retryPolicyProto = retryPolicy.ToProtobuf();

            Assert.IsNotNull(retryPolicyProto.ExponentialBackoff);

            var exponentialBackoff = retryPolicyProto.ExponentialBackoff;
            var initialBackoffProto = Duration.FromTimeSpan(initialBackoff);
            var maxBackoffProto = Duration.FromTimeSpan(maxBackoff);

            Assert.AreEqual(exponentialBackoff.Initial, initialBackoffProto);
            Assert.AreEqual(exponentialBackoff.Max, maxBackoffProto);
            Assert.AreEqual(exponentialBackoff.Multiplier, backoffMultiplier);
            Assert.AreEqual(retryPolicyProto.MaxAttempts, maxAttempts);
        }

        [TestMethod]
        public void TestInheritBackoff()
        {
            var retryPolicy = new ExponentialBackoffRetryPolicy(maxAttempts, initialBackoff, maxBackoff, backoffMultiplier);

            var initialBackoffProto = TimeSpan.FromMilliseconds(10);
            var maxBackoffProto = TimeSpan.FromSeconds(3);
            double backoffMultiplierProto = 10;

            var exponentialBackoff = new ExponentialBackoff
            {
                Initial = Duration.FromTimeSpan(initialBackoffProto),
                Max = Duration.FromTimeSpan(maxBackoffProto),
                Multiplier = (float)backoffMultiplierProto
            };
            var retryPolicyProto = new RetryPolicy
            {
                ExponentialBackoff = exponentialBackoff
            };

            var inheritedRetryPolicy = retryPolicy.InheritBackoff(retryPolicyProto);

            Assert.IsInstanceOfType(inheritedRetryPolicy, typeof(ExponentialBackoffRetryPolicy));

            var exponentialBackoffRetryPolicy = (ExponentialBackoffRetryPolicy)inheritedRetryPolicy;

            Assert.AreEqual(initialBackoffProto, exponentialBackoffRetryPolicy.InitialBackoff);
            Assert.AreEqual(maxBackoffProto, exponentialBackoffRetryPolicy.MaxBackoff);
            Assert.AreEqual(backoffMultiplierProto, exponentialBackoffRetryPolicy.BackoffMultiplier);
        }

        [TestMethod]
        [ExpectedException(typeof(InvalidOperationException))]
        public void TestInheritBackoffWithoutExponentialBackoff()
        {
            var customizedBackoff = new CustomizedBackoff();
            var retryPolicyProto = new RetryPolicy
            {
                MaxAttempts = maxAttempts,
                CustomizedBackoff = customizedBackoff
            };

            var exponentialBackoffRetryPolicy = new ExponentialBackoffRetryPolicy(maxAttempts, initialBackoff, maxBackoff, backoffMultiplier);
            exponentialBackoffRetryPolicy.InheritBackoff(retryPolicyProto);
        }
    }

}

File: php/tests/ExponentialBackoffRetryPolicyTest.php
<?php
/**
 * Licensed to the Apache Software Foundation (ASF) under one or more
 * contributor license agreements.  See the NOTICE file distributed with
 * this work for additional information regarding copyright ownership.
 * The ASF licenses this file to You under the Apache License, Version 2.0
 * (the "License"); you may not use this file except in compliance with
 * the License.  You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

namespace Apache\Rocketmq\Test;

use PHPUnit\Framework\TestCase;
require_once __DIR__ . '/../autoload.php';
require_once __DIR__ . '/../ExponentialBackoffRetryPolicy.php';

use Apache\Rocketmq\ExponentialBackoffRetryPolicy;

/**
 * Tests for ExponentialBackoffRetryPolicy.
 */
class ExponentialBackoffRetryPolicyTest extends TestCase
{
    public function testDefaultPolicy()
    {
        $policy = new ExponentialBackoffRetryPolicy();
        $this->assertEquals(3, $policy->getMaxAttempts(), "Default max attempts should be 3");

        $delay1 = $policy->getNextDelayMs(1);
        $this->assertEquals(1000, $delay1, "Attempt 1 delay should be base (1000ms)");

        $delay2 = $policy->getNextDelayMs(2);
        $this->assertEquals(2000, $delay2, "Attempt 2 delay should be 2000ms (1000 * 2)");
    }

    public function testExponentialGrowth()
    {
        $policy = new ExponentialBackoffRetryPolicy(5, 100, 10000, 2.0);

        $this->assertEquals(100, $policy->getNextDelayMs(1), "Attempt 1 should be 100ms");
        $this->assertEquals(200, $policy->getNextDelayMs(2), "Attempt 2 should be 200ms");
        $this->assertEquals(400, $policy->getNextDelayMs(3), "Attempt 3 should be 400ms");
        $this->assertEquals(800, $policy->getNextDelayMs(4), "Attempt 4 should be 800ms");
        $this->assertEquals(0, $policy->getNextDelayMs(5), "Attempt 5 should be 0 (at max)");
    }

    public function testMaxDelayCap()
    {
        $policy = new ExponentialBackoffRetryPolicy(10, 1000, 5000, 3.0);

        $delay1 = $policy->getNextDelayMs(1);
        $this->assertEquals(1000, $delay1, "Attempt 1 should be 1000ms");

        $delay2 = $policy->getNextDelayMs(2);
        $this->assertEquals(3000, $delay2, "Attempt 2 should be 3000ms");

        $delay3 = $policy->getNextDelayMs(3);
        $this->assertEquals(5000, $delay3, "Attempt 3 should be capped at 5000ms");
    }

    public function testWithJitter()
    {
        $policy = new ExponentialBackoffRetryPolicy(5, 1000, 10000, 2.0);

        $jitter = $policy->getNextDelayWithJitterMs(1);
        // Jitter range is 50%-100% of base, allow small tolerance for edge cases
        $minExpected = (int)(1000 * 0.49);
        $maxExpected = 1000;
        $this->assertTrue(
            $jitter >= $minExpected && $jitter <= $maxExpected,
            "Jitter should be between 50%% and 100%% of base ({$minExpected}-{$maxExpected}). Got: {$jitter}"
        );
    }

    public function testRejectsNegativeMaxAttempts()
    {
        $this->expectException(\InvalidArgumentException::class);
        new ExponentialBackoffRetryPolicy(0);
    }
}
